das/reader_aggregator_strategies.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"errors"
+	"math/rand"
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	"github.com/offchainlabs/nitro/arbstate"
+)
+
+var ErrNoReadersResponded = errors.New("No DAS readers responded successfully.")
+
+type aggregatorStrategy interface {
+	newInstance() aggregatorStrategyInstance
+	update([]arbstate.SimpleDASReader, map[arbstate.SimpleDASReader]readerStats)
+}
+
+type abstractAggregatorStrategy struct {
+	sync.RWMutex
+	readers []arbstate.SimpleDASReader
+	stats   map[arbstate.SimpleDASReader]readerStats
+}
+
+func (s *abstractAggregatorStrategy) update(readers []arbstate.SimpleDASReader, stats map[arbstate.SimpleDASReader]readerStats) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.readers = make([]arbstate.SimpleDASReader, len(readers))
+	copy(s.readers, readers)
+
+	s.stats = make(map[arbstate.SimpleDASReader]readerStats)
+	for k, v := range stats {
+		s.stats[k] = v
+	}
+}
+
+// Exponentially growing Explore Exploit Strategy
+type simpleExploreExploitStrategy struct {
+	iterations        uint32
+	exploreIterations uint32
+	exploitIterations uint32
+
+	abstractAggregatorStrategy
+}
+
+func (s *simpleExploreExploitStrategy) newInstance() aggregatorStrategyInstance {
+	iterations := atomic.AddUint32(&s.iterations, 1)
+
+	readerSets := make([][]arbstate.SimpleDASReader, 0)
+	s.RLock()
+	defer s.RUnlock()
+
+	readers := make([]arbstate.SimpleDASReader, len(s.readers))
+	copy(readers, s.readers)
+
+	if iterations%(s.exploreIterations+s.exploitIterations) < s.exploreIterations {
```

Comment:
Does it make sense to interleave the explore and exploit iterations, rather than having them in separate contiguous chunks?

Reply:
After discussion, we decided the current approach is best, because it allows to express interleaving reasonably well when that is wanted.